worlds/marioland2/__init__.py

```diff
@@ -0,0 +1,374 @@
+            "Pumpkin Zone 3 -> Pumpkin Zone - Secret Course 2": lambda state: state.has("Pumpkin Zone Secret 2", self.player),
+            "Pumpkin Zone 3 -> Pumpkin Zone 4": lambda state: has_level_progression(state, "Pumpkin Zone Progression", self.player, 3),
+            "Mario Zone 1 -> Mario Zone 2": lambda state: has_level_progression(state, "Mario Zone Progression", self.player),
+            "Mario Zone 2 -> Mario Zone 3": lambda state: has_level_progression(state, "Mario Zone Progression", self.player, 2),
+            "Mario Zone 3 -> Mario Zone 4": lambda state: has_level_progression(state, "Mario Zone Progression", self.player, 3),
+            "Turtle Zone 1 -> Turtle Zone 2": lambda state: has_level_progression(state, "Turtle Zone Progression", self.player),
+            "Turtle Zone 2 -> Turtle Zone - Secret Course": lambda state: state.has("Turtle Zone Secret", self.player),
+            "Turtle Zone 2 -> Turtle Zone 3": lambda state: has_level_progression(state, "Turtle Zone Progression", self.player, 2),
+            "Menu -> Mario's Castle": lambda state: ([
+                    state.has("Tree Coin", self.player), state.has("Space Coin", self.player),
+                    state.has("Macro Coin", self.player), state.has("Pumpkin Coin", self.player),
+                    state.has("Mario Coin", self.player), state.has("Turtle Coin", self.player)
+                ].count(True) >= self.options.required_golden_coins)
+        }
+        location_rules = {
+            "Hippo Zone": lambda state: state.has_any(["Hippo Bubble", "Carrot", "Swim"], self.player),
+            # It is possible, however tricky, to beat the Moon Stage without Carrot or Space Physics.
+            # However, it requires somewhat precisely jumping off enemies. Enemy shuffle may make this impossible.
+            # I have not done any testing there. Instead, I will just always make one or the other required, since
+            # it is difficult without them anyway.
+            "Space Zone 1 - Moon Stage": lambda state: state.has_any(["Space Physics", "Carrot"], self.player),
+            # One or the other is actually necessary for the secret exit.
+            "Space Zone 1 - Moon Stage Secret Exit": lambda state: state.has_any(
+                ["Space Physics", "Carrot"], self.player),
+            # Without Space Physics, you must be able to take damage once to reach the bell, and again after the bell.
+            # If bells are not shuffled, then any one powerup will do, as you can get the bell and come back.
+            # Otherwise, you need the bell item from the item pool, or you need to be able to take damage twice in one
+            # visit.
+            "Space Zone 2 - Star Stage": lambda state: has_pipe_right(state, self.player) and (state.has(
+                "Space Physics", self.player) or ((not
+                state.multiworld.worlds[self.player].options.shuffle_midway_bells) and state.has_any(["Mushroom",
```

Comment:
Any reason this can't just be:
```suggestion
                self.options.shuffle_midway_bells) and state.has_any(["Mushroom",
```

Reply:
Now moved to its own function in rules.py where there is no `self`